docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+# Upgrade Process
+
+## Abstract
+The goal of this document is to design and describe a system where upgrading a project from one version of a Package (say Angular2 `2.X`) to an ulterior version of the same package (say Angular `2.Y`, where `Y` is strictly greater than `X`), implying refactoring the code base of said project to compile and execute properly after the upgrade.
+
+The upgrade must be transitive. If the process can be applied to go from version `X` to version `X+1`, and from `X+1` to `X+2`, then the user must be able to run the same process for going from `X` to `X+2` directly, without failure.
+
+This is done by a series of steps, called Upgrade steps, that will be applied incrementally to the project. These steps are free form functions; they take a source tree and outputs a potentially different source tree, or indicates to the user what tasks are to be performed, in case such tasks cannot be done automatically.
+
+# Upgrade Step
+An upgrade step is a relation `s` between a source `A` and the smallest ideal source `B`, such that `s(A) ≡ B`. In this case, `A` would support compilation with Angular `2.X`, while `B` would support compilation with Angular `2.Y`, where `X + 1 = Y`.
+
+Put simply, a step is a function to get from one version of Angular to the next.
+
+## Properties of Upgrade Steps
+Upgrade steps must have 3 properties:
+
+* **Atomicity** ensures that each step is "all or nothing"; if one part of the step fails, the entire step should fail, and the source is left unchanged.
+* **Consistency** ensures that each step will take the source from one compilable state to another compilable state. If the state is inconsistent before the step is applied (ie. it does not compile), the step should not be applied and/or fail.
+* **Durability** ensures that once a step succeeds, it is permanent and should be committed, even in the event of a power loss. This implies that the final operation should be to commit the code to a VCS.
+
+One might figure that, adding an Isolation property, this would match the ACID principle of database transactions. Since multiple steps cannot be performed in parallel, the source is already isolated and it would be a redundant property.
+
+Please note that because of the consistency of a step, assuming the property that code compile in one version, it is imperative that code compiles in the version after the step is applied. This leads to a strong induction with regards to the "can compile" property.
+
+If we add another property; invertibility, we could technically _undo_ upgrade steps and as such, if an error is inserted in the process, we could revert the step, make the corrections, then apply the step again.
+
+### **RFC: should invertibility be a requirement or ignored.**
```

Comment:
Enforcing the durability property already means we have a copy of codebase state in a previous step. To undo a step we'd merely need to checkout the previous state.

The only case I see for invertibility is to downgrade apps, which I don't think is necessary.